Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,191 @@
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+// MARK: - ClientRequest
+
+/// Instructs the SOCKS proxy server of the target host,
+/// and how to connect.
+struct ClientRequest: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public let version: UInt8 = 5
+    
+    /// How to connect to the host.
+    public var command: Command
+    
+    /// The target host address.
+    public var addressType: AddressType
+    
+    /// Creates a new `ClientRequest`.
+    /// - parameter command: How to connect to the host.
+    /// - parameter addressType: The target host address.
+    /// - parameter desiredPort: The target host port.
+    public init(command: Command, addressType: AddressType) {
+        self.command = command
+        self.addressType = addressType
+    }
+    
+}
+
+extension ByteBuffer {
+    
+    @discardableResult mutating func writeClientRequest(_ request: ClientRequest) -> Int {
+        var written = 0
+        written += self.writeInteger(request.version)
+        written += self.writeInteger(request.command.value)
+        written += self.writeInteger(UInt8(0))
+        written += self.writeAddressType(request.addressType)
+        return written
+    }
+    
+}
+
+// MARK: - Command
+
+/// What type of connection the SOCKS server should establish with
+/// the target host.
+public struct Command: Hashable {
+    
+    /// Typically the primary connection type, suitable for HTTP.
+    public static let connect = Command(value: 0x01)
+    
+    /// Used in protocols that require the client to accept connections
+    /// from the server, e.g. FTP.
+    public static let bind = Command(value: 0x02)
+    
+    /// Used to establish an association within the UDP relay process to
+    /// handle UDP datagrams.
+    public static let udpAssociate = Command(value: 0x03)
+    
+    public var value: UInt8
+    
+    public init(value: UInt8) {
+        self.value = value
+    }
+}
+
+// MARK: - AddressType
+
+/// The address used to connect to the target host.
+public enum AddressType: Hashable {
+    
+    case address(SocketAddress)
+    
+    case domain(String, port: Int)
+    
+    /// How many bytes are needed to represent the address, excluding the port
+    var size: Int {
+        switch self {
+        case .address(.v4):
+            return 4
+        case .address(.v6):
+            return 16
+        case .address(.unixDomainSocket):
+            fatalError("Unsupported")
+        case .domain(let domain, port: _):
+            return domain.utf8.count + 1
```

Comment:
The +1 is for the null terminator, not the domain.

Reply:
My mistake... the docs even say "excluding the port"

Reply:
You've both made the point that I should document this. The +1 is because the domain type needs to have a leading byte saying how long the byte is. Unsure why they didn't just go for the NUL terminator, but here we are.